From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Henry Lin <henryl@nvidia.com>
Date: Wed, 24 Apr 2024 12:03:25 +0000
Subject: [PATCH 055/241] NVIDIA: SAUCE: phy: tegra: xusb: add vf support

In virtualization environment, xusb padctl driver does not access
clock, reset, supply regulator, and fuse calibration.

http://nvbugs/3840188

Change-Id: I076cf26bcf5cb9965e4595d0dd952683797c3c1d
Signed-off-by: Henry Lin <henryl@nvidia.com>
Reviewed-on: https://git-master.nvidia.com/r/c/linux-stable/+/3132351
(cherry picked from commit 6c5ab42bfc0d39eeb2bbdfe3609c0bab3d0c1c3d)
Signed-off-by: Bodla Rakesh Babu <rbodla@nvidia.com>
Reviewed-on: https://git-master.nvidia.com/r/c/3rdparty/canonical/linux-noble/+/3178749
GVS: buildbot_gerritrpt <buildbot_gerritrpt@nvidia.com>
Reviewed-by: Laxman Dewangan <ldewangan@nvidia.com>
(cherry picked from commit dc374ea86ae26744401eaaa513d19f37d15a5c9f)

diff --git a/drivers/phy/tegra/xusb-tegra186.c b/drivers/phy/tegra/xusb-tegra186.c
index 4452e73fb82a..1d8a9e57ecf6 100644
--- a/drivers/phy/tegra/xusb-tegra186.c
+++ b/drivers/phy/tegra/xusb-tegra186.c
@@ -609,9 +609,11 @@ static void tegra186_utmi_bias_pad_power_on(struct tegra_xusb_padctl *padctl)
 	if (!bitmap_empty(priv->utmi_pad_enabled, TEGRA_UTMI_PAD_MAX))
 		return;
 
-	err = clk_prepare_enable(priv->usb2_trk_clk);
-	if (err < 0)
-		dev_warn(dev, "failed to enable USB2 trk clock: %d\n", err);
+	if (!padctl->is_xhci_iov) {
+		err = clk_prepare_enable(priv->usb2_trk_clk);
+		if (err < 0)
+			dev_warn(dev, "failed to enable USB2 trk clock: %d\n", err);
+	}
 
 	value = padctl_readl(padctl, XUSB_PADCTL_USB2_BIAS_PAD_CTL1);
 	value &= ~USB2_TRK_START_TIMER(~0);
@@ -658,7 +660,7 @@ static void tegra186_utmi_bias_pad_power_on(struct tegra_xusb_padctl *padctl)
 		value |= USB2_TRK_HW_MODE;
 	padctl_writel(padctl, value, XUSB_PADCTL_USB2_BIAS_PAD_CTL2);
 
-	if (!padctl->soc->trk_hw_mode)
+	if (!padctl->soc->trk_hw_mode && !padctl->is_xhci_iov)
 		clk_disable_unprepare(priv->usb2_trk_clk);
 }
 
@@ -678,7 +680,8 @@ static void tegra186_utmi_bias_pad_power_off(struct tegra_xusb_padctl *padctl)
 		value = padctl_readl(padctl, XUSB_PADCTL_USB2_BIAS_PAD_CTL2);
 		value &= ~USB2_TRK_HW_MODE;
 		padctl_writel(padctl, value, XUSB_PADCTL_USB2_BIAS_PAD_CTL2);
-		clk_disable_unprepare(priv->usb2_trk_clk);
+		if (!padctl->is_xhci_iov)
+			clk_disable_unprepare(priv->usb2_trk_clk);
 	}
 
 }
@@ -910,20 +913,22 @@ static int tegra186_utmi_phy_power_on(struct phy *phy)
 	value |= TERM_SEL;
 	value &= ~HS_CURR_LEVEL(~0);
 
-	if (usb2->hs_curr_level_offset) {
-		int hs_current_level;
+	if (!padctl->is_xhci_iov) {
+		if (usb2->hs_curr_level_offset) {
+			int hs_current_level;
 
-		hs_current_level = (int)priv->calib.hs_curr_level[index] +
-						usb2->hs_curr_level_offset;
+			hs_current_level = (int)priv->calib.hs_curr_level[index] +
+							usb2->hs_curr_level_offset;
 
-		if (hs_current_level < 0)
-			hs_current_level = 0;
-		if (hs_current_level > 0x3f)
-			hs_current_level = 0x3f;
+			if (hs_current_level < 0)
+				hs_current_level = 0;
+			if (hs_current_level > 0x3f)
+				hs_current_level = 0x3f;
 
-		value |= HS_CURR_LEVEL(hs_current_level);
-	} else {
-		value |= HS_CURR_LEVEL(priv->calib.hs_curr_level[index]);
+			value |= HS_CURR_LEVEL(hs_current_level);
+		} else {
+			value |= HS_CURR_LEVEL(priv->calib.hs_curr_level[index]);
+		}
 	}
 
 	padctl_writel(padctl, value, XUSB_PADCTL_USB2_OTG_PADX_CTL0(index));
@@ -1045,11 +1050,13 @@ tegra186_usb2_pad_probe(struct tegra_xusb_padctl *padctl,
 		goto out;
 	}
 
-	priv->usb2_trk_clk = devm_clk_get(&pad->dev, "trk");
-	if (IS_ERR(priv->usb2_trk_clk)) {
-		err = PTR_ERR(priv->usb2_trk_clk);
-		dev_dbg(&pad->dev, "failed to get usb2 trk clock: %d\n", err);
-		goto unregister;
+	if (!padctl->is_xhci_iov) {
+		priv->usb2_trk_clk = devm_clk_get(&pad->dev, "trk");
+		if (IS_ERR(priv->usb2_trk_clk)) {
+			err = PTR_ERR(priv->usb2_trk_clk);
+			dev_dbg(&pad->dev, "failed to get usb2 trk clock: %d\n", err);
+			goto unregister;
+		}
 	}
 
 	err = tegra_xusb_pad_register(pad, &utmi_phy_ops);
@@ -1531,6 +1538,8 @@ tegra186_xusb_padctl_probe(struct device *dev,
 	struct tegra186_xusb_padctl *priv;
 	struct resource *res;
 	int err;
+	bool is_xhci_iov = !!of_find_property(dev->of_node, "is_xhci_iov",
+						NULL);
 
 	priv = devm_kzalloc(dev, sizeof(*priv), GFP_KERNEL);
 	if (!priv)
@@ -1544,9 +1553,11 @@ tegra186_xusb_padctl_probe(struct device *dev,
 	if (IS_ERR(priv->ao_regs))
 		return ERR_CAST(priv->ao_regs);
 
-	err = tegra186_xusb_read_fuse_calibration(priv);
-	if (err < 0)
-		return ERR_PTR(err);
+	if (!is_xhci_iov) {
+		err = tegra186_xusb_read_fuse_calibration(priv);
+		if (err < 0)
+			return ERR_PTR(err);
+	}
 
 	return &priv->base;
 }
diff --git a/drivers/phy/tegra/xusb.c b/drivers/phy/tegra/xusb.c
index c89df95aa6ca..f7bd83e5f350 100644
--- a/drivers/phy/tegra/xusb.c
+++ b/drivers/phy/tegra/xusb.c
@@ -1,6 +1,6 @@
 // SPDX-License-Identifier: GPL-2.0-only
 /*
- * Copyright (c) 2014-2022, NVIDIA CORPORATION.  All rights reserved.
+ * SPDX-FileCopyrightText: Copyright (c) 2014-2024, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
  */
 
 #include <linux/delay.h>
@@ -1196,11 +1196,15 @@ static int tegra_xusb_padctl_probe(struct platform_device *pdev)
 	if (IS_ERR(padctl))
 		return PTR_ERR(padctl);
 
+	padctl->is_xhci_iov = !!of_find_property(pdev->dev.of_node,
+						"is_xhci_iov", NULL);
+
 	platform_set_drvdata(pdev, padctl);
 	INIT_LIST_HEAD(&padctl->ports);
 	INIT_LIST_HEAD(&padctl->lanes);
 	INIT_LIST_HEAD(&padctl->pads);
 	mutex_init(&padctl->lock);
+	BLOCKING_INIT_NOTIFIER_HEAD(&padctl->notifier);
 
 	padctl->regs = devm_platform_ioremap_resource(pdev, 0);
 	if (IS_ERR(padctl->regs)) {
@@ -1208,39 +1212,41 @@ static int tegra_xusb_padctl_probe(struct platform_device *pdev)
 		goto remove;
 	}
 
-	padctl->rst = devm_reset_control_get(&pdev->dev, NULL);
-	if (IS_ERR(padctl->rst)) {
-		err = PTR_ERR(padctl->rst);
-		goto remove;
-	}
+	if (!padctl->is_xhci_iov) {
+		padctl->rst = devm_reset_control_get(&pdev->dev, NULL);
+		if (IS_ERR(padctl->rst)) {
+			err = PTR_ERR(padctl->rst);
+			goto remove;
+		}
 
-	padctl->supplies = devm_kcalloc(&pdev->dev, padctl->soc->num_supplies,
-					sizeof(*padctl->supplies), GFP_KERNEL);
-	if (!padctl->supplies) {
-		err = -ENOMEM;
-		goto remove;
-	}
+		padctl->supplies = devm_kcalloc(&pdev->dev, padctl->soc->num_supplies,
+						sizeof(*padctl->supplies), GFP_KERNEL);
+		if (!padctl->supplies) {
+			err = -ENOMEM;
+			goto remove;
+		}
 
-	regulator_bulk_set_supply_names(padctl->supplies,
-					padctl->soc->supply_names,
-					padctl->soc->num_supplies);
+		regulator_bulk_set_supply_names(padctl->supplies,
+						padctl->soc->supply_names,
+						padctl->soc->num_supplies);
 
-	err = devm_regulator_bulk_get(&pdev->dev, padctl->soc->num_supplies,
-				      padctl->supplies);
-	if (err < 0) {
-		dev_err_probe(&pdev->dev, err, "failed to get regulators\n");
-		goto remove;
-	}
+		err = devm_regulator_bulk_get(&pdev->dev, padctl->soc->num_supplies,
+					padctl->supplies);
+		if (err < 0) {
+			dev_err_probe(&pdev->dev, err, "failed to get regulators\n");
+			goto remove;
+		}
 
-	err = reset_control_deassert(padctl->rst);
-	if (err < 0)
-		goto remove;
+		err = reset_control_deassert(padctl->rst);
+		if (err < 0)
+			goto remove;
 
-	err = regulator_bulk_enable(padctl->soc->num_supplies,
-				    padctl->supplies);
-	if (err < 0) {
-		dev_err(&pdev->dev, "failed to enable supplies: %d\n", err);
-		goto reset;
+		err = regulator_bulk_enable(padctl->soc->num_supplies,
+					padctl->supplies);
+		if (err < 0) {
+			dev_err(&pdev->dev, "failed to enable supplies: %d\n", err);
+			goto reset;
+		}
 	}
 
 	err = tegra_xusb_setup_pads(padctl);
@@ -1266,9 +1272,11 @@ static int tegra_xusb_padctl_probe(struct platform_device *pdev)
 remove_pads:
 	tegra_xusb_remove_pads(padctl);
 power_down:
-	regulator_bulk_disable(padctl->soc->num_supplies, padctl->supplies);
+	if (!padctl->is_xhci_iov)
+		regulator_bulk_disable(padctl->soc->num_supplies, padctl->supplies);
 reset:
-	reset_control_assert(padctl->rst);
+	if (!padctl->is_xhci_iov)
+		reset_control_assert(padctl->rst);
 remove:
 	platform_set_drvdata(pdev, NULL);
 	soc->ops->remove(padctl);
@@ -1283,14 +1291,16 @@ static void tegra_xusb_padctl_remove(struct platform_device *pdev)
 	tegra_xusb_remove_ports(padctl);
 	tegra_xusb_remove_pads(padctl);
 
-	err = regulator_bulk_disable(padctl->soc->num_supplies,
-				     padctl->supplies);
-	if (err < 0)
-		dev_err(&pdev->dev, "failed to disable supplies: %d\n", err);
+	if (!padctl->is_xhci_iov) {
+		err = regulator_bulk_disable(padctl->soc->num_supplies,
+					padctl->supplies);
+		if (err < 0)
+			dev_err(&pdev->dev, "failed to disable supplies: %d\n", err);
 
-	err = reset_control_assert(padctl->rst);
-	if (err < 0)
-		dev_err(&pdev->dev, "failed to assert reset: %d\n", err);
+		err = reset_control_assert(padctl->rst);
+		if (err < 0)
+			dev_err(&pdev->dev, "failed to assert reset: %d\n", err);
+	}
 
 	padctl->soc->ops->remove(padctl);
 }
@@ -1371,6 +1381,20 @@ void tegra_xusb_padctl_put(struct tegra_xusb_padctl *padctl)
 }
 EXPORT_SYMBOL_GPL(tegra_xusb_padctl_put);
 
+int tegra_xusb_padctl_event_register(struct tegra_xusb_padctl *padctl,
+				     struct notifier_block *nb)
+{
+	return blocking_notifier_chain_register(&padctl->notifier, nb);
+}
+EXPORT_SYMBOL_GPL(tegra_xusb_padctl_event_register);
+
+int tegra_xusb_padctl_event_notify(struct tegra_xusb_padctl *padctl,
+				     unsigned long val)
+{
+	return blocking_notifier_call_chain(&padctl->notifier, val, padctl);
+}
+EXPORT_SYMBOL_GPL(tegra_xusb_padctl_event_notify);
+
 int tegra_xusb_padctl_usb3_save_context(struct tegra_xusb_padctl *padctl,
 					unsigned int port)
 {
diff --git a/drivers/phy/tegra/xusb.h b/drivers/phy/tegra/xusb.h
index 810b410672f3..7e7774ef2668 100644
--- a/drivers/phy/tegra/xusb.h
+++ b/drivers/phy/tegra/xusb.h
@@ -1,6 +1,6 @@
 /* SPDX-License-Identifier: GPL-2.0-only */
 /*
- * Copyright (c) 2014-2022, NVIDIA CORPORATION.  All rights reserved.
+ * SPDX-FileCopyrightText: Copyright (c) 2014-2024, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
  * Copyright (c) 2015, Google Inc.
  */
 
@@ -11,6 +11,7 @@
 #include <linux/iopoll.h>
 #include <linux/mutex.h>
 #include <linux/workqueue.h>
+#include <linux/notifier.h>
 
 #include <linux/usb/ch9.h>
 #include <linux/usb/otg.h>
@@ -458,10 +459,13 @@ struct tegra_xusb_padctl {
 	struct list_head pads;
 
 	unsigned int enable;
+	bool is_xhci_iov;
 
 	struct clk *clk;
 
 	struct regulator_bulk_data *supplies;
+
+	struct blocking_notifier_head notifier;
 };
 
 static inline void padctl_writel(struct tegra_xusb_padctl *padctl, u32 value,
diff --git a/include/linux/phy/tegra/xusb.h b/include/linux/phy/tegra/xusb.h
index 6ca51e0080ec..bcb0e9ee3d46 100644
--- a/include/linux/phy/tegra/xusb.h
+++ b/include/linux/phy/tegra/xusb.h
@@ -8,11 +8,20 @@
 
 struct tegra_xusb_padctl;
 struct device;
+struct notifier_block;
 enum usb_device_speed;
 
+#define PHY_TEGRA_XUSB_SUPPORT_EVENT
+
+#define XUSB_EVENT_VF_OFFSET		24
+#define XUSB_EVENT_VF_MASK		0xff
+
 struct tegra_xusb_padctl *tegra_xusb_padctl_get(struct device *dev);
 void tegra_xusb_padctl_put(struct tegra_xusb_padctl *padctl);
-
+int tegra_xusb_padctl_event_register(struct tegra_xusb_padctl *padctl,
+				     struct notifier_block *nb);
+int tegra_xusb_padctl_event_notify(struct tegra_xusb_padctl *padctl,
+				     unsigned long val);
 int tegra_xusb_padctl_usb3_save_context(struct tegra_xusb_padctl *padctl,
 					unsigned int port);
 int tegra_xusb_padctl_hsic_set_idle(struct tegra_xusb_padctl *padctl,
-- 
2.43.0

